Sources/SourceControl/RepositoryManager.swift

```diff
@@ -145,53 +184,93 @@ public class RepositoryManager {
+        concurrentQueue.async { 
+            // Get the handle.
+            let handle = self.getHandle(repository: repository)
+            handle.serialQueue.async {
+                let result: LookupResult
+                switch handle.status {
+                case .available:
+                    // FIXME: Need to only do this for first lookup.
+                    result = try! LookupResult {
+                        // Fetch and update the repository when it is being looked up.
+                        // We need some mechanism to maybe not do this for pinned repos.
+                        let repo = try handle.open()
+                        try repo.fetch()
+                        return handle
+                    }
+                case .pending:
+                    precondition(false, "This should never have been called")
+                    return
+                case .uninitialized, .error:
+                    // Fetch the repo.
+                    result = self.fetchRepository(of: handle)
+                    // Save the manager state.
+                    self.serialQueue.sync { 
+                        do {
+                            try self.saveState()
+                        } catch {
+                            // FIXME: Handle failure gracefully, somehow.
+                            fatalError("unable to save manager state \(error)")
+                        }
+                    }
+                }
+                completion(result)
             }
-            completion(Result(handle))
-            return
         }
-        
-        // Otherwise, fetch the repository and return a handle.
-        let subpath = RelativePath(repository.fileSystemIdentifier)
-        let handle = RepositoryHandle(manager: self, repository: repository, subpath: subpath)
-        repositories[repository.url] = handle
+    }
 
-        // Ensure nothing else exists at the subpath.
-        let repositoryPath = path.appending(subpath)
+    /// Returns the handle for repository if available, otherwise creates a new one.
+    /// Note: This method is thread safe.
+    private func getHandle(
+        repository: RepositorySpecifier
+    ) -> RepositoryHandle {
+        return serialQueue.sync {
+            // Get the handle.
+            let handle: RepositoryHandle
+            if let oldHandle = self.repositories[repository.url] {
+                handle = oldHandle
+            } else {
+                let subpath = RelativePath(repository.fileSystemIdentifier)
+                let newHandle = RepositoryHandle(manager: self, repository: repository, subpath: subpath)
+                self.repositories[repository.url] = newHandle
+                handle = newHandle
+            }
+            return handle
+        }
+    }
+
+    /// Fetches the repository of a handle in pending status and returns the result.
+    private func fetchRepository(of handle: RepositoryHandle) -> LookupResult {
```

Comment:
This method should document it is expected to be called protected by the internal queue.